File: Clients/Node/Core/src/models/audio.ts
/**
 * Audio API models for Conduit Core client library
 * Supports speech-to-text, text-to-speech, and audio translation capabilities
 */

import type { Usage } from './common';

// Audio format enums
export type AudioFormat = 
  | 'mp3' 
  | 'wav' 
  | 'flac' 
  | 'ogg' 
  | 'aac' 
  | 'opus' 
  | 'pcm' 
  | 'm4a' 
  | 'webm';

export type TranscriptionFormat = 
  | 'json' 
  | 'text' 
  | 'srt' 
  | 'vtt' 
  | 'verbose_json';

export type TimestampGranularity = 
  | 'segment' 
  | 'word';

export type TextToSpeechModel = 
  | 'tts-1' 
  | 'tts-1-hd' 
  | 'elevenlabs-tts' 
  | 'azure-tts' 
  | 'openai-tts';

export type Voice = 
  | 'alloy' 
  | 'echo' 
  | 'fable' 
  | 'onyx' 
  | 'nova' 
  | 'shimmer'
  | 'rachel'
  | 'adam'
  | 'antoni'
  | 'arnold'
  | 'josh'
  | 'sam';

export type TranscriptionModel = 
  | 'whisper-1' 
  | 'whisper-large' 
  | 'deepgram-nova' 
  | 'azure-stt' 
  | 'openai-whisper';

// Base audio interfaces
export interface AudioFile {
  /** The audio file data as Buffer, Blob, or base64 string */
  data: Buffer | Blob | string;
  /** The filename of the audio file */
  filename: string;
  /** The MIME type of the audio file */
  contentType?: string;
}

export interface VoiceSettings {
  /** Voice stability (0.0 to 1.0) */
  stability?: number;
  /** Voice similarity boost (0.0 to 1.0) */
  similarity_boost?: number;
  /** Voice style exaggeration (0.0 to 1.0) */
  style?: number;
  /** Use speaker boost for enhanced clarity */
  use_speaker_boost?: boolean;
}

// Speech-to-Text (Transcription) interfaces
export interface AudioTranscriptionRequest {
  /** The audio file to transcribe */
  file: AudioFile;
  /** The model to use for transcription */
  model: TranscriptionModel;
  /** The language of the input audio (ISO-639-1 format, e.g., 'en', 'es') */
  language?: string;
  /** An optional text to guide the model's style or continue a previous audio segment */
  prompt?: string;
  /** The format of the transcript output */
  response_format?: TranscriptionFormat;
  /** The sampling temperature (0 to 1) */
  temperature?: number;
  /** The timestamp granularities to populate for this transcription */
  timestamp_granularities?: TimestampGranularity[];
}

export interface AudioTranscriptionResponse {
  /** The transcribed text */
  text: string;
  /** The task performed (e.g., 'transcribe') */
  task?: string;
  /** The language of the input audio */
  language?: string;
  /** The duration of the input audio in seconds */
  duration?: number;
  /** Array of transcription segments with timestamps */
  segments?: TranscriptionSegment[];
  /** Array of words with timestamps (if word-level timestamps requested) */
  words?: TranscriptionWord[];
  /** Token usage information */
  usage?: Usage;
}

export interface TranscriptionSegment {
  /** Unique identifier of the segment */
  id: number;
  /** Seek offset of the segment */
  seek: number;
  /** Start time of the segment in seconds */
  start: number;
  /** End time of the segment in seconds */
  end: number;
  /** Text content of the segment */
  text: string;
  /** Array of token IDs for the text content */
  tokens: number[];
  /** Temperature parameter used for generation */
  temperature: number;
  /** Average logprob of the segment */
  avg_logprob: number;
  /** Compression ratio of the segment */
  compression_ratio: number;
  /** Probability of no speech */
  no_speech_prob: number;
}

export interface TranscriptionWord {
  /** The text content of the word */
  word: string;
  /** Start time of the word in seconds */
  start: number;
  /** End time of the word in seconds */
  end: number;
}

// Audio Translation interfaces
export interface AudioTranslationRequest {
  /** The audio file to translate */
  file: AudioFile;
  /** The model to use for translation */
  model: TranscriptionModel;
  /** An optional text to guide the model's style or continue a previous audio segment */
  prompt?: string;
  /** The format of the transcript output */
  response_format?: TranscriptionFormat;
  /** The sampling temperature (0 to 1) */
  temperature?: number;
}

export interface AudioTranslationResponse {
  /** The translated text (always in English) */
  text: string;
  /** The task performed (e.g., 'translate') */
  task?: string;
  /** The language of the input audio */
  language?: string;
  /** The duration of the input audio in seconds */
  duration?: number;
  /** Array of translation segments with timestamps */
  segments?: TranscriptionSegment[];
  /** Array of words with timestamps (if word-level timestamps requested) */
  words?: TranscriptionWord[];
  /** Token usage information */
  usage?: Usage;
}

// Text-to-Speech interfaces
export interface TextToSpeechRequest {
  /** The model to use for speech generation */
  model: TextToSpeechModel;
  /** The text to convert to speech (max 4096 characters) */
  input: string;
  /** The voice to use for speech generation */
  voice: Voice;
  /** The format to audio in */
  response_format?: AudioFormat;
  /** The speed of the generated audio (0.25 to 4.0) */
  speed?: number;
  /** Advanced voice settings (for compatible providers) */
  voice_settings?: VoiceSettings;
}

export interface TextToSpeechResponse {
  /** The generated audio data as Buffer */
  audio: Buffer;
  /** The format of the returned audio */
  format: AudioFormat;
  /** Additional metadata about the generation */
  metadata?: {
    /** Duration of the generated audio in seconds */
    duration?: number;
    /** Size of the audio data in bytes */
    size?: number;
    /** Sample rate of the audio */
    sample_rate?: number;
    /** Number of audio channels */
    channels?: number;
  };
  /** Token usage information (if applicable) */
  usage?: Usage;
}

// Hybrid Audio interfaces (STT + LLM + TTS pipeline)
export interface HybridAudioRequest {
  /** The input audio file for processing */
  file: AudioFile;
  /** The model configuration for each stage */
  models: {
    /** Speech-to-text model */
    transcription: TranscriptionModel;
    /** Chat completion model for LLM processing */
    chat: string;
    /** Text-to-speech model */
    speech: TextToSpeechModel;
  };
  /** Voice configuration for TTS output */
  voice: Voice;
  /** System prompt for the LLM stage */
  system_prompt?: string;
  /** Additional context for the conversation */
  context?: string;
  /** Language settings */
  language?: string;
  /** Temperature settings for each stage */
  temperature?: {
    transcription?: number;
    chat?: number;
  };
  /** Voice settings for TTS */
  voice_settings?: VoiceSettings;
  /** Session ID for conversation continuity */
  session_id?: string;
}

export interface HybridAudioResponse {
  /** The generated audio response */
  audio: Buffer;
  /** The transcribed input text */
  transcription: string;
  /** The LLM's text response */
  llm_response: string;
  /** Metadata for each processing stage */
  stages: {
    transcription: {
      duration: number;
      confidence?: number;
      language?: string;
    };
    llm: {
      duration: number;
      tokens_used: number;
      model_used: string;
    };
    speech: {
      duration: number;
      audio_duration: number;
      format: AudioFormat;
    };
  };
  /** Combined usage statistics */
  usage: {
    transcription_tokens?: number;
    llm_tokens: Usage;
    total_processing_time_ms: number;
  };
  /** Session information */
  session_id?: string;
}

// Real-time Audio interfaces (for WebSocket connections)
export interface RealtimeConnectionRequest {
  /** The model to use for real-time processing */
  model: string;
  /** Voice configuration for real-time TTS */
  voice?: Voice;
  /** Audio format for input/output */
  audio_format?: AudioFormat;
  /** Sample rate for audio processing */
  sample_rate?: number;
  /** Whether to enable voice activity detection */
  enable_vad?: boolean;
  /** Session configuration */
  session_config?: RealtimeSessionConfig;
}

export interface RealtimeSessionConfig {
  /** Instructions for the assistant */
  instructions?: string;
  /** Audio input/output configuration */
  input_audio_format?: AudioFormat;
  output_audio_format?: AudioFormat;
  /** Voice configuration */
  voice?: Voice;
  /** Model configuration */
  model?: string;
  /** Temperature for responses */
  temperature?: number;
  /** Maximum response tokens */
  max_response_output_tokens?: number;
  /** Tools available to the assistant */
  tools?: Array<{
    type: 'function';
    name: string;
    description?: string;
    parameters?: Record<string, unknown>;
  }>;
  /** Turn detection configuration */
  turn_detection?: {
    type: 'server_vad' | 'none';
    threshold?: number;
    prefix_padding_ms?: number;
    silence_duration_ms?: number;
  };
}

export interface RealtimeMessage {
  /** The type of real-time message */
  type: 
    | 'session.created'
    | 'session.updated' 
    | 'input_audio_buffer.append'
    | 'input_audio_buffer.commit'
    | 'input_audio_buffer.clear'
    | 'conversation.item.create'
    | 'response.create'
    | 'response.cancel'
    | 'error';
  /** The message data */
  data?: Record<string, unknown>;
  /** Message ID for tracking */
  event_id?: string;
}

export interface RealtimeSession {
  /** Unique session identifier */
  id: string;
  /** Current session status */
  status: 'active' | 'idle' | 'ended' | 'error';
  /** Session configuration */
  config: RealtimeSessionConfig;
  /** Connection metadata */
  connection: {
    created_at: string;
    last_activity: string;
    duration_seconds: number;
  };
  /** Usage statistics for the session */
  usage: {
    total_audio_minutes: number;
    total_tokens: number;
    input_tokens: number;
    output_tokens: number;
  };
}

// Audio utility types
export interface AudioMetadata {
  /** Duration in seconds */
  duration: number;
  /** File size in bytes */
  size: number;
  /** Audio format */
  format: AudioFormat;
  /** Sample rate in Hz */
  sample_rate: number;
  /** Number of channels */
  channels: number;
  /** Bit depth */
  bit_depth?: number;
  /** Bitrate in kbps */
  bitrate?: number;
}

export interface AudioProcessingOptions {
  /** Maximum file size in bytes (default: 25MB) */
  max_file_size?: number;
  /** Supported audio formats */
  supported_formats?: AudioFormat[];
  /** Quality settings */
  quality?: 'low' | 'medium' | 'high' | 'ultra';
  /** Whether to normalize audio */
  normalize?: boolean;
  /** Whether to remove noise */
  denoise?: boolean;
}

// Error types specific to audio processing
export interface AudioError {
  code: 
    | 'invalid_audio_format'
    | 'file_too_large'
    | 'audio_too_long'
    | 'unsupported_language'
    | 'transcription_failed'
    | 'synthesis_failed'
    | 'realtime_connection_failed';
  message: string;
  details?: Record<string, unknown>;
}

// Audio validation utilities
export interface AudioValidation {
  /** Validate audio file format and size */
  validateAudioFile(file: AudioFile, options?: AudioProcessingOptions): Promise<boolean>;
  /** Get audio metadata */
  getAudioMetadata(file: AudioFile): Promise<AudioMetadata>;
  /** Convert audio format */
  convertAudioFormat(file: AudioFile, targetFormat: AudioFormat): Promise<AudioFile>;
}

// Types are exported automatically by TypeScript

File: archive/Clients.DotNet/ConduitLLM.CoreClient/src/Models/Audio.cs
using System.ComponentModel.DataAnnotations;
using System.Text.Json.Serialization;

namespace ConduitLLM.CoreClient.Models;

/// <summary>
/// Supported audio formats for input and output.
/// </summary>
public enum AudioFormat
{
    /// <summary>MP3 format.</summary>
    Mp3,
    /// <summary>WAV format.</summary>
    Wav,
    /// <summary>FLAC format.</summary>
    Flac,
    /// <summary>OGG format.</summary>
    Ogg,
    /// <summary>AAC format.</summary>
    Aac,
    /// <summary>Opus format.</summary>
    Opus,
    /// <summary>PCM format.</summary>
    Pcm,
    /// <summary>M4A format.</summary>
    M4a,
    /// <summary>WebM format.</summary>
    Webm
}

/// <summary>
/// Supported transcription output formats.
/// </summary>
public enum TranscriptionFormat
{
    /// <summary>JSON format.</summary>
    Json,
    /// <summary>Plain text format.</summary>
    Text,
    /// <summary>SRT subtitle format.</summary>
    Srt,
    /// <summary>VTT subtitle format.</summary>
    Vtt,
    /// <summary>Verbose JSON format with timestamps.</summary>
    VerboseJson
}

/// <summary>
/// Timestamp granularity options for transcriptions.
/// </summary>
public enum TimestampGranularity
{
    /// <summary>Segment-level timestamps.</summary>
    Segment,
    /// <summary>Word-level timestamps.</summary>
    Word
}

/// <summary>
/// Available text-to-speech models.
/// </summary>
public enum TextToSpeechModel
{
    /// <summary>OpenAI TTS-1 model.</summary>
    Tts1,
    /// <summary>OpenAI TTS-1-HD model.</summary>
    Tts1Hd,
    /// <summary>ElevenLabs TTS model.</summary>
    ElevenlabsTts,
    /// <summary>Azure TTS model.</summary>
    AzureTts,
    /// <summary>OpenAI TTS model.</summary>
    OpenaiTts
}

/// <summary>
/// Available voices for text-to-speech generation.
/// </summary>
public enum Voice
{
    /// <summary>Alloy voice.</summary>
    Alloy,
    /// <summary>Echo voice.</summary>
    Echo,
    /// <summary>Fable voice.</summary>
    Fable,
    /// <summary>Onyx voice.</summary>
    Onyx,
    /// <summary>Nova voice.</summary>
    Nova,
    /// <summary>Shimmer voice.</summary>
    Shimmer,
    /// <summary>Rachel voice.</summary>
    Rachel,
    /// <summary>Adam voice.</summary>
    Adam,
    /// <summary>Antoni voice.</summary>
    Antoni,
    /// <summary>Arnold voice.</summary>
    Arnold,
    /// <summary>Josh voice.</summary>
    Josh,
    /// <summary>Sam voice.</summary>
    Sam
}

/// <summary>
/// Available transcription models.
/// </summary>
public enum TranscriptionModel
{
    /// <summary>OpenAI Whisper-1 model.</summary>
    Whisper1,
    /// <summary>OpenAI Whisper-Large model.</summary>
    WhisperLarge,
    /// <summary>Deepgram Nova model.</summary>
    DeepgramNova,
    /// <summary>Azure Speech-to-Text model.</summary>
    AzureStt,
    /// <summary>OpenAI Whisper model.</summary>
    OpenaiWhisper
}

/// <summary>
/// Represents an audio file for processing.
/// </summary>
public class AudioFile
{
    /// <summary>
    /// Gets or sets the audio file data.
    /// </summary>
    [Required]
    public byte[] Data { get; set; } = Array.Empty<byte>();

    /// <summary>
    /// Gets or sets the filename of the audio file.
    /// </summary>
    [Required]
    public string Filename { get; set; } = string.Empty;

    /// <summary>
    /// Gets or sets the MIME type of the audio file.
    /// </summary>
    public string? ContentType { get; set; }
}

/// <summary>
/// Advanced voice settings for text-to-speech generation.
/// </summary>
public class VoiceSettings
{
    /// <summary>
    /// Gets or sets the voice stability (0.0 to 1.0).
    /// </summary>
    [Range(0.0, 1.0)]
    public double? Stability { get; set; }

    /// <summary>
    /// Gets or sets the voice similarity boost (0.0 to 1.0).
    /// </summary>
    [Range(0.0, 1.0)]
    [JsonPropertyName("similarity_boost")]
    public double? SimilarityBoost { get; set; }

    /// <summary>
    /// Gets or sets the voice style exaggeration (0.0 to 1.0).
    /// </summary>
    [Range(0.0, 1.0)]
    public double? Style { get; set; }

    /// <summary>
    /// Gets or sets whether to use speaker boost for enhanced clarity.
    /// </summary>
    [JsonPropertyName("use_speaker_boost")]
    public bool? UseSpeakerBoost { get; set; }
}

/// <summary>
/// Request for audio transcription (speech-to-text).
/// </summary>
public class AudioTranscriptionRequest
{
    /// <summary>
    /// Gets or sets the audio file to transcribe.
    /// </summary>
    [Required]
    public AudioFile File { get; set; } = new();

    /// <summary>
    /// Gets or sets the model to use for transcription.
    /// </summary>
    [Required]
    public TranscriptionModel Model { get; set; }

    /// <summary>
    /// Gets or sets the language of the input audio (ISO-639-1 format).
    /// </summary>
    public string? Language { get; set; }

    /// <summary>
    /// Gets or sets an optional text to guide the model's style.
    /// </summary>
    public string? Prompt { get; set; }

    /// <summary>
    /// Gets or sets the format of the transcript output.
    /// </summary>
    [JsonPropertyName("response_format")]
    public TranscriptionFormat? ResponseFormat { get; set; }

    /// <summary>
    /// Gets or sets the sampling temperature (0 to 1).
    /// </summary>
    [Range(0.0, 1.0)]
    public double? Temperature { get; set; }

    /// <summary>
    /// Gets or sets the timestamp granularities to populate.
    /// </summary>
    [JsonPropertyName("timestamp_granularities")]
    public TimestampGranularity[]? TimestampGranularities { get; set; }
}

/// <summary>
/// Response from audio transcription.
/// </summary>
public class AudioTranscriptionResponse
{
    /// <summary>
    /// Gets or sets the transcribed text.
    /// </summary>
    [Required]
    public string Text { get; set; } = string.Empty;

    /// <summary>
    /// Gets or sets the task performed.
    /// </summary>
    public string? Task { get; set; }

    /// <summary>
    /// Gets or sets the language of the input audio.
    /// </summary>
    public string? Language { get; set; }

    /// <summary>
    /// Gets or sets the duration of the input audio in seconds.
    /// </summary>
    public double? Duration { get; set; }

    /// <summary>
    /// Gets or sets the transcription segments with timestamps.
    /// </summary>
    public TranscriptionSegment[]? Segments { get; set; }

    /// <summary>
    /// Gets or sets the words with timestamps.
    /// </summary>
    public TranscriptionWord[]? Words { get; set; }

    /// <summary>
    /// Gets or sets the token usage information.
    /// </summary>
    public Usage? Usage { get; set; }
}

/// <summary>
/// Represents a segment of transcribed audio with timing information.
/// </summary>
public class TranscriptionSegment
{
    /// <summary>
    /// Gets or sets the unique identifier of the segment.
    /// </summary>
    public int Id { get; set; }

    /// <summary>
    /// Gets or sets the seek offset of the segment.
    /// </summary>
    public int Seek { get; set; }

    /// <summary>
    /// Gets or sets the start time of the segment in seconds.
    /// </summary>
    public double Start { get; set; }

    /// <summary>
    /// Gets or sets the end time of the segment in seconds.
    /// </summary>
    public double End { get; set; }

    /// <summary>
    /// Gets or sets the text content of the segment.
    /// </summary>
    public string Text { get; set; } = string.Empty;

    /// <summary>
    /// Gets or sets the token IDs for the text content.
    /// </summary>
    public int[] Tokens { get; set; } = Array.Empty<int>();

    /// <summary>
    /// Gets or sets the temperature parameter used for generation.
    /// </summary>
    public double Temperature { get; set; }

    /// <summary>
    /// Gets or sets the average log probability of the segment.
    /// </summary>
    [JsonPropertyName("avg_logprob")]
    public double AvgLogprob { get; set; }

    /// <summary>
    /// Gets or sets the compression ratio of the segment.
    /// </summary>
    [JsonPropertyName("compression_ratio")]
    public double CompressionRatio { get; set; }

    /// <summary>
    /// Gets or sets the probability of no speech.
    /// </summary>
    [JsonPropertyName("no_speech_prob")]
    public double NoSpeechProb { get; set; }
}

/// <summary>
/// Represents a word in transcribed audio with timing information.
/// </summary>
public class TranscriptionWord
{
    /// <summary>
    /// Gets or sets the text content of the word.
    /// </summary>
    public string Word { get; set; } = string.Empty;

    /// <summary>
    /// Gets or sets the start time of the word in seconds.
    /// </summary>
    public double Start { get; set; }

    /// <summary>
    /// Gets or sets the end time of the word in seconds.
    /// </summary>
    public double End { get; set; }
}

/// <summary>
/// Request for audio translation to English.
/// </summary>
public class AudioTranslationRequest
{
    /// <summary>
    /// Gets or sets the audio file to translate.
    /// </summary>
    [Required]
    public AudioFile File { get; set; } = new();

    /// <summary>
    /// Gets or sets the model to use for translation.
    /// </summary>
    [Required]
    public TranscriptionModel Model { get; set; }

    /// <summary>
    /// Gets or sets an optional text to guide the model's style.
    /// </summary>
    public string? Prompt { get; set; }

    /// <summary>
    /// Gets or sets the format of the transcript output.
    /// </summary>
    [JsonPropertyName("response_format")]
    public TranscriptionFormat? ResponseFormat { get; set; }

    /// <summary>
    /// Gets or sets the sampling temperature (0 to 1).
    /// </summary>
    [Range(0.0, 1.0)]
    public double? Temperature { get; set; }
}

/// <summary>
/// Response from audio translation.
/// </summary>
public class AudioTranslationResponse
{
    /// <summary>
    /// Gets or sets the translated text (always in English).
    /// </summary>
    [Required]
    public string Text { get; set; } = string.Empty;

    /// <summary>
    /// Gets or sets the task performed.
    /// </summary>
    public string? Task { get; set; }

    /// <summary>
    /// Gets or sets the language of the input audio.
    /// </summary>
    public string? Language { get; set; }

    /// <summary>
    /// Gets or sets the duration of the input audio in seconds.
    /// </summary>
    public double? Duration { get; set; }

    /// <summary>
    /// Gets or sets the translation segments with timestamps.
    /// </summary>
    public TranscriptionSegment[]? Segments { get; set; }

    /// <summary>
    /// Gets or sets the words with timestamps.
    /// </summary>
    public TranscriptionWord[]? Words { get; set; }

    /// <summary>
    /// Gets or sets the token usage information.
    /// </summary>
    public Usage? Usage { get; set; }
}

/// <summary>
/// Request for text-to-speech generation.
/// </summary>
public class TextToSpeechRequest
{
    /// <summary>
    /// Gets or sets the model to use for speech generation.
    /// </summary>
    [Required]
    public TextToSpeechModel Model { get; set; }

    /// <summary>
    /// Gets or sets the text to convert to speech (max 4096 characters).
    /// </summary>
    [Required]
    [MaxLength(4096)]
    public string Input { get; set; } = string.Empty;

    /// <summary>
    /// Gets or sets the voice to use for speech generation.
    /// </summary>
    [Required]
    public Voice Voice { get; set; }

    /// <summary>
    /// Gets or sets the format to audio in.
    /// </summary>
    [JsonPropertyName("response_format")]
    public AudioFormat? ResponseFormat { get; set; }

    /// <summary>
    /// Gets or sets the speed of the generated audio (0.25 to 4.0).
    /// </summary>
    [Range(0.25, 4.0)]
    public double? Speed { get; set; }

    /// <summary>
    /// Gets or sets advanced voice settings.
    /// </summary>
    [JsonPropertyName("voice_settings")]
    public VoiceSettings? VoiceSettings { get; set; }
}

/// <summary>
/// Response from text-to-speech generation.
/// </summary>
public class TextToSpeechResponse
{
    /// <summary>
    /// Gets or sets the generated audio data.
    /// </summary>
    [Required]
    public byte[] Audio { get; set; } = Array.Empty<byte>();

    /// <summary>
    /// Gets or sets the format of the returned audio.
    /// </summary>
    public AudioFormat Format { get; set; }

    /// <summary>
    /// Gets or sets additional metadata about the generation.
    /// </summary>
    public AudioMetadata? Metadata { get; set; }

    /// <summary>
    /// Gets or sets token usage information.
    /// </summary>
    public Usage? Usage { get; set; }
}

/// <summary>
/// Metadata about generated or processed audio.
/// </summary>
public class AudioMetadata
{
    /// <summary>
    /// Gets or sets the duration of the audio in seconds.
    /// </summary>
    public double? Duration { get; set; }

    /// <summary>
    /// Gets or sets the size of the audio data in bytes.
    /// </summary>
    public long? Size { get; set; }

    /// <summary>
    /// Gets or sets the sample rate of the audio.
    /// </summary>
    [JsonPropertyName("sample_rate")]
    public int? SampleRate { get; set; }

    /// <summary>
    /// Gets or sets the number of audio channels.
    /// </summary>
    public int? Channels { get; set; }

    /// <summary>
    /// Gets or sets the bit depth of the audio.
    /// </summary>
    [JsonPropertyName("bit_depth")]
    public int? BitDepth { get; set; }

    /// <summary>
    /// Gets or sets the bitrate in kbps.
    /// </summary>
    public int? Bitrate { get; set; }
}

/// <summary>
/// Configuration for hybrid audio processing models.
/// </summary>
public class HybridAudioModels
{
    /// <summary>
    /// Gets or sets the speech-to-text model.
    /// </summary>
    [Required]
    public TranscriptionModel Transcription { get; set; }

    /// <summary>
    /// Gets or sets the chat completion model for LLM processing.
    /// </summary>
    [Required]
    public string Chat { get; set; } = string.Empty;

    /// <summary>
    /// Gets or sets the text-to-speech model.
    /// </summary>
    [Required]
    public TextToSpeechModel Speech { get; set; }
}

/// <summary>
/// Temperature settings for different stages of hybrid processing.
/// </summary>
public class HybridTemperatureSettings
{
    /// <summary>
    /// Gets or sets the temperature for transcription.
    /// </summary>
    public double? Transcription { get; set; }

    /// <summary>
    /// Gets or sets the temperature for chat completion.
    /// </summary>
    public double? Chat { get; set; }
}

/// <summary>
/// Request for hybrid audio processing (STT + LLM + TTS pipeline).
/// </summary>
public class HybridAudioRequest
{
    /// <summary>
    /// Gets or sets the input audio file for processing.
    /// </summary>
    [Required]
    public AudioFile File { get; set; } = new();

    /// <summary>
    /// Gets or sets the model configuration for each stage.
    /// </summary>
    [Required]
    public HybridAudioModels Models { get; set; } = new();

    /// <summary>
    /// Gets or sets the voice configuration for TTS output.
    /// </summary>
    [Required]
    public Voice Voice { get; set; }

    /// <summary>
    /// Gets or sets the system prompt for the LLM stage.
    /// </summary>
    [JsonPropertyName("system_prompt")]
    public string? SystemPrompt { get; set; }

    /// <summary>
    /// Gets or sets additional context for the conversation.
    /// </summary>
    public string? Context { get; set; }

    /// <summary>
    /// Gets or sets the language settings.
    /// </summary>
    public string? Language { get; set; }

    /// <summary>
    /// Gets or sets temperature settings for each stage.
    /// </summary>
    public HybridTemperatureSettings? Temperature { get; set; }

    /// <summary>
    /// Gets or sets voice settings for TTS.
    /// </summary>
    [JsonPropertyName("voice_settings")]
    public VoiceSettings? VoiceSettings { get; set; }

    /// <summary>
    /// Gets or sets the session ID for conversation continuity.
    /// </summary>
    [JsonPropertyName("session_id")]
    public string? SessionId { get; set; }
}

/// <summary>
/// Metadata for processing stages in hybrid audio.
/// </summary>
public class HybridStageMetadata
{
    /// <summary>
    /// Gets or sets transcription stage metadata.
    /// </summary>
    public TranscriptionStageMetadata Transcription { get; set; } = new();

    /// <summary>
    /// Gets or sets LLM stage metadata.
    /// </summary>
    public LlmStageMetadata Llm { get; set; } = new();

    /// <summary>
    /// Gets or sets speech generation stage metadata.
    /// </summary>
    public SpeechStageMetadata Speech { get; set; } = new();
}

/// <summary>
/// Metadata for the transcription stage.
/// </summary>
public class TranscriptionStageMetadata
{
    /// <summary>
    /// Gets or sets the processing duration in milliseconds.
    /// </summary>
    public double Duration { get; set; }

    /// <summary>
    /// Gets or sets the transcription confidence.
    /// </summary>
    public double? Confidence { get; set; }

    /// <summary>
    /// Gets or sets the detected language.
    /// </summary>
    public string? Language { get; set; }
}

/// <summary>
/// Metadata for the LLM stage.
/// </summary>
public class LlmStageMetadata
{
    /// <summary>
    /// Gets or sets the processing duration in milliseconds.
    /// </summary>
    public double Duration { get; set; }

    /// <summary>
    /// Gets or sets the number of tokens used.
    /// </summary>
    [JsonPropertyName("tokens_used")]
    public int TokensUsed { get; set; }

    /// <summary>
    /// Gets or sets the model used.
    /// </summary>
    [JsonPropertyName("model_used")]
    public string ModelUsed { get; set; } = string.Empty;
}

/// <summary>
/// Metadata for the speech generation stage.
/// </summary>
public class SpeechStageMetadata
{
    /// <summary>
    /// Gets or sets the processing duration in milliseconds.
    /// </summary>
    public double Duration { get; set; }

    /// <summary>
    /// Gets or sets the duration of the generated audio in seconds.
    /// </summary>
    [JsonPropertyName("audio_duration")]
    public double AudioDuration { get; set; }

    /// <summary>
    /// Gets or sets the format of the generated audio.
    /// </summary>
    public AudioFormat Format { get; set; }
}

/// <summary>
/// Usage statistics for hybrid audio processing.
/// </summary>
public class HybridAudioUsage
{
    /// <summary>
    /// Gets or sets transcription tokens used.
    /// </summary>
    [JsonPropertyName("transcription_tokens")]
    public int? TranscriptionTokens { get; set; }

    /// <summary>
    /// Gets or sets LLM token usage.
    /// </summary>
    [JsonPropertyName("llm_tokens")]
    public Usage LlmTokens { get; set; } = new();

    /// <summary>
    /// Gets or sets total processing time in milliseconds.
    /// </summary>
    [JsonPropertyName("total_processing_time_ms")]
    public double TotalProcessingTimeMs { get; set; }
}

/// <summary>
/// Response from hybrid audio processing.
/// </summary>
public class HybridAudioResponse
{
    /// <summary>
    /// Gets or sets the generated audio response.
    /// </summary>
    [Required]
    public byte[] Audio { get; set; } = Array.Empty<byte>();

    /// <summary>
    /// Gets or sets the transcribed input text.
    /// </summary>
    [Required]
    public string Transcription { get; set; } = string.Empty;

    /// <summary>
    /// Gets or sets the LLM's text response.
    /// </summary>
    [JsonPropertyName("llm_response")]
    [Required]
    public string LlmResponse { get; set; } = string.Empty;

    /// <summary>
    /// Gets or sets metadata for each processing stage.
    /// </summary>
    public HybridStageMetadata Stages { get; set; } = new();

    /// <summary>
    /// Gets or sets combined usage statistics.
    /// </summary>
    public HybridAudioUsage Usage { get; set; } = new();

    /// <summary>
    /// Gets or sets session information.
    /// </summary>
    [JsonPropertyName("session_id")]
    public string? SessionId { get; set; }
}

/// <summary>
/// Audio processing options and constraints.
/// </summary>
public class AudioProcessingOptions
{
    /// <summary>
    /// Gets or sets the maximum file size in bytes (default: 25MB).
    /// </summary>
    [JsonPropertyName("max_file_size")]
    public long MaxFileSize { get; set; } = 25 * 1024 * 1024;

    /// <summary>
    /// Gets or sets the supported audio formats.
    /// </summary>
    [JsonPropertyName("supported_formats")]
    public AudioFormat[] SupportedFormats { get; set; } = 
    {
        AudioFormat.Mp3, AudioFormat.Wav, AudioFormat.Flac, 
        AudioFormat.Ogg, AudioFormat.Aac, AudioFormat.Opus
    };

    /// <summary>
    /// Gets or sets the quality settings.
    /// </summary>
    public string Quality { get; set; } = "medium";

    /// <summary>
    /// Gets or sets whether to normalize audio.
    /// </summary>
    public bool Normalize { get; set; } = false;

    /// <summary>
    /// Gets or sets whether to remove noise.
    /// </summary>
    public bool Denoise { get; set; } = false;
}

/// <summary>
/// Audio-specific error information.
/// </summary>
public class AudioError
{
    /// <summary>
    /// Gets or sets the error code.
    /// </summary>
    public string Code { get; set; } = string.Empty;

    /// <summary>
    /// Gets or sets the error message.
    /// </summary>
    public string Message { get; set; } = string.Empty;

    /// <summary>
    /// Gets or sets additional error details.
    /// </summary>
    public Dictionary<string, object>? Details { get; set; }
}